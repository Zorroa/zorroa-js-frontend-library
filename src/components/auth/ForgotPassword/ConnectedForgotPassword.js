import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { forgotPassword, authError } from '../../../actions/authAction'
import ForgotPassword from './ForgotPassword'

const ConnectedForgotPassword = connect(
  state => ({
    user: state.auth.user,
    error: state.auth.error,
    passwordResetStatus: state.auth.passwordResetRequestStatus,
  }),
  dispatch => ({
    actions: bindActionCreators({ forgotPassword, authError }, dispatch),
  }),
)(ForgotPassword)

export default withRouter(ConnectedForgotPassword)
