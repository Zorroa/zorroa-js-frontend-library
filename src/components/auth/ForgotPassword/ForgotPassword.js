import PropTypes from 'prop-types'
import React, { Component } from 'react'
import Logo from '../../Logo'
import Heading from '../../Heading'
import Paragraph from '../../Paragraph'
import FlashMessage from '../../FlashMessage'
import Button from '../../Form/Button'
import User from '../../../models/User'
import { isValidEmail } from '../../../services/jsUtil'

export default class ForgotPassword extends Component {
  static propTypes = {
    user: PropTypes.instanceOf(User),
    actions: PropTypes.shape({
      forgotPassword: PropTypes.func.isRequired,
      authError: PropTypes.func.isRequired,
    }),
    passwordResetStatus: PropTypes.string.isRequired,
    history: PropTypes.shape({
      goBack: PropTypes.func.isRequired,
      push: PropTypes.func.isRequired,
    }),
  }

  state = {
    email: '',
    showSuccessMessage: false,
  }

  static get contextTypes() {
    return {
      router: PropTypes.object,
    }
  }

  componentWillMount() {
    this.props.actions.authError() // clear any existing errors
  }

  componentDidUpdate(prevProps) {
    if (this.shouldUpdateSuccessStatus(prevProps)) {
      this.updateResetStatusState()
    }
  }

  shouldUpdateSuccessStatus({ passwordResetStatus }) {
    return passwordResetStatus !== this.props.passwordResetStatus
  }

  updateResetStatusState() {
    this.setState({
      showSuccessMessage: this.props.passwordResetStatus === 'succeeded',
    })
  }

  changeEmail = event => {
    this.setState({ email: event.target.value })
  }

  cancel = event => {
    this.props.history.goBack()
  }

  submit = event => {
    event.preventDefault()
    this.props.actions.forgotPassword(this.state.email)
  }

  login = event => {
    event.preventDefault()
    this.props.history.push('/signin')
  }

  render() {
    const { passwordResetStatus } = this.props
    const { email, showSuccessMessage } = this.state
    const disabled = !email || !email.length || !isValidEmail(email)

    return (
      <div className="auth">
        <div className="auth-box">
          <div className="auth-logo">
            <Logo />
          </div>
          <div className="auth-form">
            {showSuccessMessage === true && (
              <div className="auth-confirm">
                <Heading>Thank you.</Heading>
                <Paragraph>
                  If that email address matches an exisiting account you will
                  recieve a password reset link.
                </Paragraph>
                <div className="auth-button-group">
                  <Button onClick={this.login}>Login</Button>
                </div>
              </div>
            )}
            {showSuccessMessage === false && (
              <form onSubmit={!disabled && this.submit}>
                <div className="auth-error">
                  {passwordResetStatus === 'errored' && (
                    <FlashMessage look="information">
                      There was a problem sending the request. Please try again
                      in a few minutes.
                    </FlashMessage>
                  )}
                </div>
                <div className="auth-help">
                  <Heading>Let’s get your account back</Heading>
                  <Paragraph>
                    Enter your email address and we’ll send you a link to reset
                    your password.
                  </Paragraph>
                </div>
                <div className="auth-field">
                  <input
                    className="auth-input"
                    type="text"
                    value={email}
                    onChange={this.changeEmail}
                  />
                  <label className="auth-label">Email Address</label>
                </div>
                <div className="auth-button-group">
                  <Button
                    type="submit"
                    state={
                      passwordResetStatus === 'pending' ? 'loading' : undefined
                    }
                    disabled={disabled}>
                    Send Password Link
                  </Button>
                </div>
              </form>
            )}
          </div>
        </div>
      </div>
    )
  }
}
