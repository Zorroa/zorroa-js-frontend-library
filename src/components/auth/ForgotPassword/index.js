import ForgotPassword from './ConnectedForgotPassword'
export default ForgotPassword
