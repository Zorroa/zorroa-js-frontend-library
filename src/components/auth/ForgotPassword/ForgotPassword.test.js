/* eslint-env jest */
import React from 'react'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-15'
import ForgotPassword from './ForgotPassword'
import renderer from 'react-test-renderer'
jest.mock('../../Logo')
configure({ adapter: new Adapter() })

function generateActions() {
  const forgotPassword = jest.fn()
  const authError = jest.fn()
  return {
    forgotPassword,
    authError,
  }
}

function generateProps(specialProps) {
  return {
    passwordResetStatus: '',
    history: {
      goBack: jest.fn(),
      push: jest.fn(),
    },
    actions: generateActions(),
    ...specialProps,
  }
}

describe('<ForgotPassword />', () => {
  describe('changeEmail()', () => {
    it('Should update the email', () => {
      const props = generateProps()
      const component = shallow(<ForgotPassword {...props} />)
      component.instance().changeEmail({
        target: {
          value: 'abc@zorroa.ai',
        },
      })
      expect(component.state('email')).toEqual('abc@zorroa.ai')
    })
  })

  describe('When the reset request has not been sent', () => {
    it('Should render the form', () => {
      const props = generateProps()
      const tree = renderer.create(<ForgotPassword {...props} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('When the reset request is submitted sucesfully', () => {
    it('Should render a success message', () => {
      const props = generateProps()
      const tree = renderer.create(<ForgotPassword {...props} />)

      tree.getInstance().setState({
        showSuccessMessage: true,
      })

      expect(tree.toJSON()).toMatchSnapshot()
    })
  })

  describe('When there is a server error', () => {
    it('Should display a friendly error message', () => {
      const props = generateProps({
        passwordResetStatus: 'errored',
      })
      const tree = renderer.create(<ForgotPassword {...props} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
