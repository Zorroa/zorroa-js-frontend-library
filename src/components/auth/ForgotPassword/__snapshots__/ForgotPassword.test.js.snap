// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForgotPassword /> When the reset request has not been sent Should render the form 1`] = `
<div
  className="auth"
>
  <div
    className="auth-box"
  >
    <div
      className="auth-logo"
    />
    <div
      className="auth-form"
    >
      <form
        onSubmit={false}
      >
        <div
          className="auth-error"
        />
        <div
          className="auth-help"
        >
          <h1
            className="Heading Heading--large"
          >
            Let’s get your account back
          </h1>
          <p
            className="Paragraph Paragraph--normal"
          >
            Enter your email address and we’ll send you a link to reset your password.
          </p>
        </div>
        <div
          className="auth-field"
        >
          <input
            className="auth-input"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            className="auth-label"
          >
            Email Address
          </label>
        </div>
        <div
          className="auth-button-group"
        >
          <button
            className="FormButton FormButton--disabled"
            disabled={true}
            type="submit"
          >
            <span
              className="FormButton__state FormButton__state--inactive"
            />
            <span
              className="FormButton__label"
            >
              Send Password Link
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`<ForgotPassword /> When the reset request is submitted sucesfully Should render a success message 1`] = `
<div
  className="auth"
>
  <div
    className="auth-box"
  >
    <div
      className="auth-logo"
    />
    <div
      className="auth-form"
    >
      <div
        className="auth-confirm"
      >
        <h1
          className="Heading Heading--large"
        >
          Thank you.
        </h1>
        <p
          className="Paragraph Paragraph--normal"
        >
          If that email address matches an exisiting account you will recieve a password reset link.
        </p>
        <div
          className="auth-button-group"
        >
          <button
            className="FormButton"
            onClick={[Function]}
            type="button"
          >
            <span
              className="FormButton__state FormButton__state--inactive"
            />
            <span
              className="FormButton__label"
            >
              Login
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ForgotPassword /> When there is a server error Should display a friendly error message 1`] = `
<div
  className="auth"
>
  <div
    className="auth-box"
  >
    <div
      className="auth-logo"
    />
    <div
      className="auth-form"
    >
      <form
        onSubmit={false}
      >
        <div
          className="auth-error"
        >
          <div
            className="FlashMessage"
          >
            <div
              className="FlashMessage__body FlashMessage__body--info"
            >
              There was a problem sending the request. Please try again in a few minutes.
            </div>
          </div>
        </div>
        <div
          className="auth-help"
        >
          <h1
            className="Heading Heading--large"
          >
            Let’s get your account back
          </h1>
          <p
            className="Paragraph Paragraph--normal"
          >
            Enter your email address and we’ll send you a link to reset your password.
          </p>
        </div>
        <div
          className="auth-field"
        >
          <input
            className="auth-input"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            className="auth-label"
          >
            Email Address
          </label>
        </div>
        <div
          className="auth-button-group"
        >
          <button
            className="FormButton FormButton--disabled"
            disabled={true}
            type="submit"
          >
            <span
              className="FormButton__state FormButton__state--inactive"
            />
            <span
              className="FormButton__label"
            >
              Send Password Link
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
